src/time/location.cr

```diff
@@ -0,0 +1,295 @@
+      # TODO: Maybe reduce to Int16 (+/- 18 hours).
+      raise InvalidTimezoneOffsetError.new(offset) if offset >= SECONDS_PER_DAY || offset <= -SECONDS_PER_DAY
+    end
+
+    def inspect(io : IO)
+      io << "Time::Zone<"
+      io << offset
+      io << ", " << name
+      io << " (DST)" if dst?
+      io << '>'
+    end
+  end
+
+  # :nodoc:
+  record ZoneTransition, when : Int64, index : UInt8, standard : Bool, utc : Bool do
+    getter? standard, utc
+
+    def inspect(io : IO)
+      io << "Time::ZoneTransition<"
+      io << '#' << index << ", "
+      Time.epoch(self.when).to_s("%F %T", io)
+      io << ", STD" if standard?
+      io << ", UTC" if utc?
+      io << '>'
+    end
+  end
+
+  # Describes the Coordinated Universal Time (UTC).
+  UTC = new "UTC", [Zone::UTC]
+
+  property name : String
+  property zones : Array(Zone)
+
+  # Most lookups will be for the current time.
+  # To avoid the binary search through tx, keep a
+  # static one-element cache that gives the correct
+  # zone for the time when the Location was created.
+  # The units for @cached_range are seconds
+  # since January 1, 1970 UTC, to match the argument
+  # to `#lookup`.
+  @cached_range : Tuple(Int64, Int64)
+  @cached_zone : Zone
+
+  # Creates a `Location` instance named *name* with fixed *offset*.
+  def self.fixed(name : String, offset : Int32)
+    new name, [Zone.new(name, offset, false)]
+  end
+
+  # Creates a `Location` instance with fixed *offset*.
+  def self.fixed(offset : Int32)
+    span = offset.abs.seconds
+    name = sprintf("%s%02d:%02d", offset.sign < 0 ? '-' : '+', span.hours, span.minutes)
+    fixed name, offset
+  end
+
+  # Returns the `Location` with the given name.
+  #
+  # This uses a list of paths to look for timezone data. Each path can
+  # either point to a directory or an uncompressed ZIP file.
+  # System-specific default paths are provided by the implementation.
+  #
+  # The first timezone data matching the given name that is successfully loaded
+  # and parsed is returned.
+  # A custom lookup path can be set as environment variable `ZONEFILE`.
+  #
+  # Special names:
+  # * `"UTC"` and empty string `""` return `Location::UTC`
+  # * `"Local"` returns `Location.local`
+  #
+  # This method caches files based on the modification time, so subsequent loads
+  # of the same location name will return the same instance of `Location` unless
+  # the timezone database has been updated in between.
+  #
+  # Example:
+  # `ZONEFILE=/path/to/zoneinfo.zip crystal eval 'pp Location.load("Custom/Location")'`
+  def self.load(name : String) : Location
+    case name
+    when "", "UTC"
+      UTC
+    when "Local"
+      local
+    when .includes?(".."), .starts_with?('/'), .starts_with?('\\')
+      # No valid IANA Time Zone name contains a single dot,
+      # much less dot dot. Likewise, none begin with a slash.
+      raise InvalidLocationNameError.new(name)
+    else
+      if zoneinfo = ENV["ZONEINFO"]?
```

Comment:
Shouldn't be `ZONEFILE` here?

Reply:
I followed Go's lead here and I don't think `ZONEFILE` makes much sense because it could be either a folder or a (ZIP) file.

Reply:
Oh, that's fine, but the docs should say `ZONEINFO` then 😄

Reply:
Hehe, true